### PR TITLE
feat(types): ContractMeta

### PR DESCRIPTION
**Pull request by turadg**
_incidental_

## Description

In adding a term to a contract I forgot to update its `customTermsShape`. The type checker should detect that. 

This makes it do so.

### Security Considerations
n/a

### Scaling Considerations
n/a

### Documentation Considerations
n/a

### Testing Considerations
I tested some failures but didn't add a unit test. Not opposed.

### Upgrade Considerations

n/a

### CODE REVIEW

File: packages/agoric-cli/test/upgrade-contract/fixed-contract.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { AmountMath } from '@agoric/ertp';
 import { makeDurableZone } from '@agoric/zone/durable.js';
 
-/** @type {ContractMeta} */
+/** @type {ContractMeta<typeof start>} */
```
Comment (dckc) [inline packages/agoric-cli/test/upgrade-contract/fixed-contract.js:4]: ideally, we'd have a type test showing that this is not a breaking change that forces all clients to change.

one alternative would be to leave the parameter implicit here.

File: packages/zoe/src/contractFacet/types-ambient.d.ts [hunk 2]
```diff
@@ -217,9 +217,15 @@ type OfferHandler<OR extends unknown = unknown, OA = never> =
...
+type ContractMeta<
+  SF extends // import inline to maintain ambient mode
+    import('../zoeService/utils').ContractStartFunction = import('../zoeService/utils').ContractStartFunction,
+> = {
+  customTermsShape?: Record<
+    Parameters<SF>[0] extends ZCF<infer CT> ? keyof CT : never,
+    Pattern
+  >;
+  privateArgsShape?: Record<keyof Parameters<SF>[1], Pattern>;
   upgradability?: 'none' | 'canBeUpgraded' | 'canUpgrade' | undefined;
```
Comment (dckc) [inline packages/zoe/src/contractFacet/types-ambient.d.ts:234]: aside: What's the difference between `canBeUpgraded` and `canUpgrade`?

While working on [contract upgrade docs](https://docs.agoric.com/guides/zoe/contract-upgrade.html), I couldn't find any definitions of these... not even any distinctions made in code.

File: packages/inter-protocol/src/provisionPool.js [hunk 3]
```diff
@@ -16,7 +16,7 @@ import { prepareProvisionPoolKit } from './provisionPoolKit.js';
 
 /** @import {Marshal} from '@endo/marshal'; */
 
-/** @type {ContractMeta} */
+/** @type {ContractMeta<typeof start>} */
 export const meta = {
```
Comment (dckc) [inline packages/inter-protocol/src/provisionPool.js]: Aside: Is the upgrade of zcf that knows to look for `privateArgsShape` under `meta` deployed to mainnet?

@Chris-Hibbert is working on upgrading the provisionPool in an upcoming release.

**Review by dckc — APPROVED** [review]

File: packages/agoric-cli/test/upgrade-contract/fixed-contract.js (hunk 1 above)
Comment (turadg) [reply packages/agoric-cli/test/upgrade-contract/fixed-contract.js:4]: agreed. I've reverted a few contracts to not take the generic.

File: packages/zoe/src/contractFacet/types-ambient.d.ts (hunk 2 above)
Comment (turadg) [reply packages/zoe/src/contractFacet/types-ambient.d.ts:234]: good question. I had to read through the code myself to refresh. I added docs on the type

File: packages/inter-protocol/src/provisionPool.js (hunk 3 above)
Comment (turadg) [reply packages/inter-protocol/src/provisionPool.js]: > Is the upgrade … deployed

Part of upcoming upgrade-16.  https://github.com/Agoric/agoric-sdk/blob/dev-upgrade-16/golang/cosmos/app/app.go#L899

**Pull request merged**